### PR TITLE
Add transceiver info CLI support to show output from TRANSCEIVER_INFO for ZR

### PR DESCRIPTION
**Pull request by mihirpat1**
Signed-off-by: Mihir Patel <patelmi@microsoft.com>

#### What I did
Added support for "show interfaces transceiver info" CLI

#### How I did it
The "show interfaces transceiver info" has been built on top "sfpshow info" script.
Overall, this CLI has been planned to replace "show interface transceiver eeprom" CLI in future. With this code change, the output of "show interface transceiver eeprom" will change for CMIS based transceivers wherein CMIS specific fields as described in the HLD (https://github.com/sonic-net/SONiC/blob/master/doc/platform_api/CMIS_and_C-CMIS_support_for_ZR.md#211-transceiver-info-table) will be displayed for CMIS based transceivers.
The output of  "show interface transceiver eeprom" will remain unchanged for non CMIS based transceivers.

#### How to verify it
Following tests were performed:
1.	"sfpshow info" o/p for VendorA 400ZR transceiver
2.	"sfpshow info" o/p for VendorB 400ZR transceiver
3.	"sfpshow info" o/p for 100G transceiver
4.	"sfpshow info" o/p for 10G transceiver
5.	"sfpshow info" o/p for all ports
6.	"sfpshow info" o/p for multi-asic switch (400ZR transceiver)
7.	"sfpshow info" o/p for multi-asic switch (100G transceiver)
8.	"sfpshow info" o/p for multi-asic switch (for all ports)
9.	"show interfaces transceiver info" o/p for VendorA 400ZR transceiver
10.	"show interfaces transceiver info" o/p for VendorB 400ZR transceiver
11.	"show interfaces transceiver info" o/p for 100G transceiver
12.	"show interfaces transceiver info" o/p for 10G transceiver
13.	"show interfaces transceiver info" o/p for multi-asic switch (400ZR transceiver)
14.	"show interfaces transceiver info" o/p for multi-asic switch (100G transceiver)
15.	"show interfaces transceiver info" o/p for multi-asic switch (for all ports)
16. "sfpshow info" o/p for QSFP-DD transceiver (non-ZR)
17. "show interfaces transceiver info" o/p for QSFP-DD transceiver (non-ZR)
18. "show interfaces transceiver info" o/p on Dell z9332f

#### Previous command output (if the output of a command-line utility has changed)
Output before change for CMIS based transceiver:
```
root@sonic:/home/admin# sfpshow eeprom -p Ethernet0
Ethernet0: SFP EEPROM detected
        Application Advertisement: 400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, DWDM, amplified - Media Assign (0x1)
                                   400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, Single Wavelength, Unamplified - Media Assign (0x1)
                                   100GAUI-2 C2M (Annex 135G) - Host Assign (0x55) - 400ZR, DWDM, amplified - Media Assign (0x1)
        Connector: LC
        Encoding: N/A
        Extended Identifier: Power Class 8 (20.0W Max)
        Extended RateSelect Compliance: N/A
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Length Cable Assembly(m): 0.0
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: sm_media_interface
        Vendor Date Code(YYYY-MM-DD Lot): 2021-11-19   
        Vendor Name: XXXX
        Vendor OUI: xx-xx-xx
        Vendor PN: XXX
        Vendor Rev: XXX
        Vendor SN: 0123456789

root@sonic:/home/admin# 

root@sonic:/home/admin# show interfaces transceiver eeprom Ethernet0
Ethernet0: SFP EEPROM detected
        Application Advertisement: 400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, DWDM, amplified - Media Assign (0x1)
                                   400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, Single Wavelength, Unamplified - Media Assign (0x1)
                                   100GAUI-2 C2M (Annex 135G) - Host Assign (0x55) - 400ZR, DWDM, amplified - Media Assign (0x1)
        Connector: LC
        Encoding: N/A
        Extended Identifier: Power Class 8 (20.0W Max)
        Extended RateSelect Compliance: N/A
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Length Cable Assembly(m): 0.0
        Nominal Bit Rate(100Mbs): 0
        Specification compliance: sm_media_interface
        Vendor Date Code(YYYY-MM-DD Lot): 2021-11-19   
        Vendor Name: XXXX
        Vendor OUI: xx-xx-xx
        Vendor PN: XXX
        Vendor Rev: XXX
        Vendor SN: 0123456789
root@sonic:/home/admin# 

```

#### New command output (if the output of a command-line utility has changed)
Output after change for CMIS based transceiver:
```
root@sonic:/home/admin# sfpshow eeprom -p Ethernet0
Ethernet0: SFP EEPROM detected
        Active Firmware: X.X
        Active application selected code assigned to host lane 1: 1
        Active application selected code assigned to host lane 2: 1
        Active application selected code assigned to host lane 3: 1
        Active application selected code assigned to host lane 4: 1
        Active application selected code assigned to host lane 5: 1
        Active application selected code assigned to host lane 6: 1
        Active application selected code assigned to host lane 7: 1
        Active application selected code assigned to host lane 8: 1
        Application Advertisement: 400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, DWDM, amplified - Media Assign (0x1)
                                   400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, Single Wavelength, Unamplified - Media Assign (0x1)
                                   100GAUI-2 C2M (Annex 135G) - Host Assign (0x55) - 400ZR, DWDM, amplified - Media Assign (0x1)
        CMIS Rev: 4.1
        Connector: LC
        Encoding: N/A
        Extended Identifier: Power Class 8 (20.0W Max)
        Extended RateSelect Compliance: N/A
        Host Lane Count: 8
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Inactive Firmware: X.X
        Length Cable Assembly(m): 0.0
        Media Interface Technology: 1550 nm DFB
        Media Lane Count: 1
        Module Hardware Rev: X.X
        Nominal Bit Rate(100Mbs): 0
        Specification Compliance: sm_media_interface
        Supported Max Laser Frequency: 196100
        Supported Max TX Power: 4.0
        Supported Min Laser Frequency: 191300
        Supported Min TX Power: -22.9
        Vendor Date Code(YYYY-MM-DD Lot): 2021-11-19   
        Vendor Name: XXXX
        Vendor OUI: xx-xx-xx
        Vendor PN: XXX
        Vendor Rev: XXX
        Vendor SN: 0123456789       

root@sonic:/home/admin# show interfaces transceiver eeprom Ethernet0
Ethernet0: SFP EEPROM detected
        Active Firmware: X.X
        Active application selected code assigned to host lane 1: 1
        Active application selected code assigned to host lane 2: 1
        Active application selected code assigned to host lane 3: 1
        Active application selected code assigned to host lane 4: 1
        Active application selected code assigned to host lane 5: 1
        Active application selected code assigned to host lane 6: 1
        Active application selected code assigned to host lane 7: 1
        Active application selected code assigned to host lane 8: 1
        Application Advertisement: 400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, DWDM, amplified - Media Assign (0x1)
                                   400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, Single Wavelength, Unamplified - Media Assign (0x1)
                                   100GAUI-2 C2M (Annex 135G) - Host Assign (0x55) - 400ZR, DWDM, amplified - Media Assign (0x1)
        CMIS Rev: 4.1
        Connector: LC
        Encoding: N/A
        Extended Identifier: Power Class 8 (20.0W Max)
        Extended RateSelect Compliance: N/A
        Host Lane Count: 8
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Inactive Firmware: X.X
        Length Cable Assembly(m): 0.0
        Media Interface Technology: 1550 nm DFB
        Media Lane Count: 1
        Module Hardware Rev: X.X
        Nominal Bit Rate(100Mbs): 0
        Specification Compliance: sm_media_interface
        Supported Max Laser Frequency: 196100
        Supported Max TX Power: 4.0
        Supported Min Laser Frequency: 191300
        Supported Min TX Power: -22.9
        Vendor Date Code(YYYY-MM-DD Lot): 2021-11-19   
        Vendor Name: XXXX
        Vendor OUI: xx-xx-xx
        Vendor PN: XXX
        Vendor Rev: XXX
        Vendor SN: 0123456789       
root@sonic:/home/admin# 
```
**New CLI output ("show interfaces transceiver info \<interface\>")**
Output for 400ZR
```
root@sonic:/home/admin# show interfaces transceiver info Ethernet0
Ethernet0: SFP EEPROM detected
        Active Firmware: X.X
        Active application selected code assigned to host lane 1: 1
        Active application selected code assigned to host lane 2: 1
        Active application selected code assigned to host lane 3: 1
        Active application selected code assigned to host lane 4: 1
        Active application selected code assigned to host lane 5: 1
        Active application selected code assigned to host lane 6: 1
        Active application selected code assigned to host lane 7: 1
        Active application selected code assigned to host lane 8: 1
        Application Advertisement: 400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, DWDM, amplified - Media Assign (0x1)
                                   400GAUI-8 C2M (Annex 120E) - Host Assign (0x1) - 400ZR, Single Wavelength, Unamplified - Media Assign (0x1)
                                   100GAUI-2 C2M (Annex 135G) - Host Assign (0x55) - 400ZR, DWDM, amplified - Media Assign (0x1)
        CMIS Rev: 4.1
        Connector: LC
        Encoding: N/A
        Extended Identifier: Power Class 8 (20.0W Max)
        Extended RateSelect Compliance: N/A
        Host Lane Count: 8
        Identifier: QSFP-DD Double Density 8X Pluggable Transceiver
        Inactive Firmware: X.X
        Length Cable Assembly(m): 0.0
        Media Interface Technology: 1550 nm DFB
        Media Lane Count: 1
        Module Hardware Rev: X.X
        Nominal Bit Rate(100Mbs): 0
        Specification Compliance: sm_media_interface
        Supported Max Laser Frequency: 196100
        Supported Max TX Power: 4.0
        Supported Min Laser Frequency: 191300
        Supported Min TX Power: -22.9
        Vendor Date Code(YYYY-MM-DD Lot): 2021-11-19   
        Vendor Name: XXXX
        Vendor OUI: xx-xx-xx
        Vendor PN: XXX
        Vendor Rev: XXX
        Vendor SN: 0123456789       
root@sonic:/home/admin# 
```

Output for 100G transceiver
```
root@sonic:/home/admin# show interfaces transceiver info Ethernet224
Ethernet224: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: LC
        Encoding: 64B/66B
        Extended Identifier: Power Class 6 Module (4.5W max.), No CLEI code present in Page 02h, CDR present in TX, CDR present in RX
        Extended RateSelect Compliance: Unknown
        Identifier: QSFP28 or later
        Length(km): 1.0
        Nominal Bit Rate(100Mbs): 255
        Specification compliance:
                10/40G Ethernet Compliance Code: Extended
                Extended Specification Compliance: 100GBASE-DR
                Fibre Channel Link Length: Unknown
                Fibre Channel Speed: Unknown
                Fibre Channel Transmission Media: Unknown
                Fibre Channel Transmitter Technology: Unknown
                Gigabit Ethernet Compliant Codes: Unknown
                SAS/SATA Compliance Codes: Unknown
                SONET Compliance Codes: Unknown
        Vendor Date Code(YYYY-MM-DD Lot): 2021-02-22   
        Vendor Name: XXXX
        Vendor OUI: xx-xx-xx
        Vendor PN: XXX
        Vendor Rev: XXX
        Vendor SN: 0123456789       
root@sonic:/home/admin# 
```

### CODE REVIEW

**Comment by mihirpat1** [comment]
@prgeor @shyam77git @jaganbal-a @longhuan-cisco
It will be great if you can help in reviewing this

**Comment by jaganbal-a** [comment]
@mihirpat1 
Now with this changes, show int trans eeprom and show int trans info will display the same info for a 100G/ 400G modules?

**Comment by mihirpat1** [comment]
> @mihirpat1 Now with this changes, show int trans eeprom and show int trans info will display the same info for a 100G/ 400G modules?

Yes, that is correct.

**Comment by longhuan-cisco** [comment]
In the PR description, looks like, the "New command output" section still shows the old CLI format:
```
root@sonic:/home/admin# show interfaces transceiver eeprom Ethernet0
```
I didn't see new CLI format "show interfaces transceiver info" in it.

**Comment by mihirpat1** [comment]
> In the PR description, looks like, the "New command output" section still shows the old CLI format:
> 
> ```
> root@sonic:/home/admin# show interfaces transceiver eeprom Ethernet0
> ```
> 
> I didn't see new CLI format "show interfaces transceiver info" in it.

@longhuan-cisco I have added it now.

File: utilities_common/sfp_helper.py [hunk 1]
```diff
@@ -19,6 +19,43 @@
     'application_advertisement': 'Application Advertisement'
 }
 
+CMIS_DATA_MAP = {
```
Comment (longhuan-cisco) [inline utilities_common/sfp_helper.py:22]: CMIS_DATA_MAP seems to be just a super set of QSFP_DATA_MAP. 
Would it be better to only list the CMIS specific data here (the delta), then later get the entire cmis_data_map by doing QSFP_DATA_MAP.update(CMIS_SPECIFIC_DATA_MAP)?  (the naming for CMIS_SPECIFC_DATA_MAP might not be perfect, you may choose whatever naming you feel suited it better)

File: scripts/sfpshow [hunk 2]
```diff
@@ -280,8 +280,26 @@ class SFPShow(object):
                                                 units)
         return output
 
-    # Convert sfp info in DB to cli output string
-    def convert_sfp_info_to_output_string(self, sfp_info_dict):
+    # Convert cmis sfp info in DB to cli output string
+    def convert_cmis_sfp_info_to_output_string(self, sfp_info_dict):
```
Comment (longhuan-cisco) [inline scripts/sfpshow:284]: Looks like, convert_cmis_sfp_info_to_output_string and convert_non_cmis_sfp_info_to_output_string have quite some code in common.
would it better to combine them into one function to reuse the code more?
something like:
```
    def convert_sfp_info_to_output_string(self, sfp_info_dict):
        indent = ' ' * 8
        output = ''
        
        data_map = CMIS_DATA_MAP if is_cmis else QSFP_DATA_MAP
        sorted_data_map_keys = sorted(data_map, key= data_map.get)
        for key in sorted_data_map_keys:
            if key == 'cable_type':
                output += '{}{}: {}\n'.format(indent, sfp_info_dict['cable_type'], sfp_info_dict['cable_length'])
            elif key == 'cable_length':
                pass
            elif key == 'specification_compliance':
                if not is_cmis:
                     ....
                else:
                     ...
            elif key == 'application_advertisement':
                output += covert_application_advertisement_to_output_string(indent, sfp_info_dict)
            else:
                output += '{}{}: {}\n'.format(indent, data_map[key], sfp_info_dict[key])

        return output
```

File: tests/sfp_test.py [hunk 3]
```diff
@@ -545,6 +591,12 @@ def test_qsfp_dd_pm_with_ns(self):
         expected = "Ethernet0: Transceiver performance monitoring not applicable"
         assert result_lines == expected
 
+    def test_cmis_sfp_info_with_ns(self):
+        runner = CliRunner()
+        result = runner.invoke(show.cli.commands["interfaces"].commands["transceiver"].commands["info"], ["Ethernet48 -n asic0"])
+        assert result.exit_code == 0
+        assert "\n".join([ l.rstrip() for l in result.output.split('\n')]) == test_cmis_eeprom_output
+
     def test_sfp_eeprom_all(self):
```
[the remark this answers is not in the document]
Comment (mihirpat1) [reply tests/sfp_test.py:640]: Added a testcase now.

File: utilities_common/sfp_helper.py (hunk 1 above)
Comment (mihirpat1) [reply utilities_common/sfp_helper.py]: Yes, have modified based on this now.

File: scripts/sfpshow (hunk 2 above)
Comment (mihirpat1) [reply scripts/sfpshow]: Yes, addressed this now

**Review by longhuan-cisco — APPROVED** [review]

**Comment by mihirpat1** [comment]
@keboliu, can you please help in reviewing this

**Comment by prgeor** [comment]
@mihirpat1 currently we don't have way to know via CLI to show what is the configured laser frequency and Tx power for 400G ZR? Can we show these two values in the "transceiver info" output? Or you have any other idea?

1 Configured Laser Frequency
2. Configured Tx Power

File: scripts/sfpshow [hunk 4]
```diff
@@ -587,6 +589,23 @@ def eeprom(port, dump_dom, namespace):
     sfp.get_eeprom()
     sfp.display_eeprom()
 
+# 'eeprom' subcommand
```
Comment (prgeor) [inline scripts/sfpshow:592]: i thought this "info" does not come under as subcommand of 'eeprom' since we will remove 'eeprom' command in future?

Comment (mihirpat1) [reply scripts/sfpshow]: Yes, this is a typo in comment. I have corrected it now.

**Comment by mihirpat1** [comment]
> @mihirpat1 currently we don't have way to know via CLI to show what is the configured laser frequency and Tx power for 400G ZR? Can we show these two values in the "transceiver info" output? Or you have any other idea?
> 
> 1 Configured Laser Frequency 2. Configured Tx Power

We are planning to display this as part of "show interface transceiver dom" as laser_config_freq and tx_config_power are currently part of TRANSCEIVER_DOM_SENSOR state DB.

**Comment by keboliu** [comment]
Questions:
1. The output of "show interface transceiver eeprom" and "show interface transceiver info" will be equal right?
2. What is the motivation to use the new command to replace the current one? "info" will be more accurate than "eeprom"? 
3. If I understood the code correctly, seems the new command will not dump the dom info,  if in the future we want to deprecate the old one, we need "sfpshow info" to have the same capability, right?

**Comment by mihirpat1** [comment]
Please find my response inline

> Questions:
> 
> 1. The output of "show interface transceiver eeprom" and "show interface transceiver info" will be equal right?
**[MP] Yes, the output of both the CLIs will be the same**

> 2. What is the motivation to use the new command to replace the current one? "info" is more accurate than "eeprom"?
**[MP] As part of discussing the work items for 400ZR optics, we decided to replace "show interfaces transceiver eeprom" CLI with "show interfaces transceiver info" to make it inline to the TRANSCEIVER_INFO in STATE_DB. All "show interfaces transceiver XXX" CLIs have been planned to be made in line with their respective "TRANSCEIVER_TABLE" name in STATE_DB. Hence, "show interfaces transceiver eeprom" CLI will be deprecated in future.**

> 3. If I understood the code correctly, seems the new command will not dump the dom info,  if in the future we want to deprecate the old one, we need "sfpshow info" to have the same capability, right?
**[MP] For dumping the DOM sensor and DOM threshold information, we are planning to add support for "show interfaces transceiver dom" CLI soon. This CLI will dump information from TRANSCEIVER_DOM_SENSOR and TRANSCEIVER_DOM_THRESHOLD STATE_DB tables. Hence, we have removed the option to dump DOM related informations using "show interfaces transceiver info" CLI**

**Review by longhuan-cisco — APPROVED** [review]

**Review by keboliu — APPROVED** [review]

**Review by prgeor — APPROVED** [review]

**Comment by prgeor** [comment]
@StormLiangMS @yxieca please help cherry pick to your branches. labels added

**Comment by yxieca** [comment]
@mihirpat1 please create separate PR for 202205 branch.

**Comment by yxieca** [comment]
@mihirpat1 can you create separate PR for 202205 branch for this change?

**Comment by mihirpat1** [comment]
> @mihirpat1 can you create separate PR for 202205 branch for this change?

@yxieca - I am currently waiting for @kannankvs to resolve the git conflict in Command-Reference.md file since the file is currently corrupted. Once this is fix, cherry pick should likely work automatically since the current conflict is with just Command-Reference.md.

**Comment by yxieca** [comment]
> > @mihirpat1 can you create separate PR for 202205 branch for this change?
> 
> @yxieca - I am currently waiting for @kannankvs to resolve the git conflict in Command-Reference.md file since the file is currently corrupted. Once this is fix, cherry pick should likely work automatically since the current conflict is with just Command-Reference.md.

@mihirpat1 if you are watching the other PR. pleae ping me when the conflict is resolved. I'll try again by then. Thanks.

**Comment by mihirpat1** [comment]
> > > @mihirpat1 can you create separate PR for 202205 branch for this change?
> > 
> > 
> > @yxieca - I am currently waiting for @kannankvs to resolve the git conflict in Command-Reference.md file since the file is currently corrupted. Once this is fix, cherry pick should likely work automatically since the current conflict is with just Command-Reference.md.
> 
> @mihirpat1 if you are watching the other PR. pleae ping me when the conflict is resolved. I'll try again by then. Thanks.

@yxieca Sure, will ping you once the conflict is resolved. @kannankvs - please share the PR link with us once you upload the fix

**Comment by yxieca** [comment]
@mihirpat1 there is still cherry-pick conflict. maybe you need to create separate PR for 202205 branch?

**Comment by mihirpat1** [comment]
> @mihirpat1 there is still cherry-pick conflict. maybe you need to create separate PR for 202205 branch?

@yxieca, sure will raise a separate PR once @kannankvs merges one of the commits which was missed with the previous merge on 202205.

**Pull request merged**